Draft and default update (#59)
ivpk/uapi: draft/openapi/docs.md

components:

1. **Data Agent** reads data directly from **Data source** or through a **Data
   Service**. Data Agent supports multiple data exchange protocols and formats.

   Data Agent uses DSA table to transform data between physical and logical
   data models and between source protocol and format into standardized UAPI
   data exchange protocol using JSON format.

2. **Data Agent** performs data authorization, using authorization scopes.
   Authorization scopes are validated with **Auth Service**. Using
   authorization scopes and DSA, Data Agent applies data filters in order to
   return only those data objects, which are available for a specific data
   user.

3. **API Gateway** works as a reverse proxy and load balancer, accepts requests
   from data users, validates these requests against OpenAPI schemas received
   from **Data Catalog**, validates authorization tokens with **Auth service**,
   provides request logging, rate limiting and other services.

   **API Gateway** after initial validation, passed requests to **Data Agent**,
   which reads data from primary sources and provides protocol and format
   conversions.

> 4. **Data user** authenticates using **Auth service** and receives access token,

Reviewer: ```suggestion
4. **Client** authenticates using **Auth service** and receives access token,
```

Siūlyčiau naudoti OAuth/OIDC sąvokas. OIDC kontekste, "User" turi kitokią prasmę.

Taip pat siūlyčiau įtraukti ir "End-User" ar "User" sąvoką, kuri yra labai svarbu UDTS kontekste.

Author: Pataisyta į Client. Kokia būtų "End-User" ar "User" sąvokos prasmė, ką reikia įtraukti?

Reviewer: Tokia pati prasmė, kaip yra apibrėžta OAuth specifikacijoje:

> **resource owner**
>
> An entity capable of granting access to a protected resource.
> When the resource owner is a person, it is referred to as an
> end-user.

Ir OIDC specifikacijoje:

> **End-User**
>
> Human participant.
>
> **Entity**
>
> Something that has a separate and distinct existence and that can be identified in a context. An End-User is one example of an Entity.

UDTS specifikacijoje, User dalyvauja User authorization flow, kur yra User componentas, kuris neatsispindi architektūroje.

Author: Reikia atnaujinti diagramas (įkelti editable svg) ir peržiūrėti data/metadata flow aprašymus, referuoti į client/user auth kituose skyriuose. Diagramose nuimti vartų ir auth serverio hardcodintus domain. Taip pat peržiūrėti Auth ir ten išskirti auth domain ir pagal `/.well-known/openid-configuration` surasti kokį token path naudoja spinta, kad pateikti pavyzdyje.

   which authorization scopes. Using access token, Data user makes request to
   API Gateway by providing access token and receives data or error response if

>    something goes wrong.

Reviewer: Siūlyčiau įtraukti ir "User" dalį, kadangi "Autorization" aprašo User flow, bet architektūros schemoje, tai niekaip neatsispindi.

# Concepts

> ## API URL structure

Reviewer: Skyrius "Concepts" apibrėžia esmines sąvokas. Todėl manau, kad "API URL structure" turėtu būti pateiktas atskirai, gal būt kaip top-level skyrius arba prie Architecture?

Tas pats ir su kirais skyriais "Concepts" viduje:

- Semantic interpretation of API URLs
- URI and Semantic Resource Identification Overview

Concepts skyriuje siūlau tiesiog pateikti tokias sąvokas, arba gal net skyrelius "Identifier" skyriaus viduje?:

- Identifier
  - URL - UDTS URL struktūros paaiškinimas.
  - URI - JSON-LT `@id` globalus identifikatorius, ryšys su KDSA.

Ir papildomai galima pridėti naujai atsiradusias sąvokas, kurių nebuvo:

- Publisher (org)
- Information System
- Data Service
- Scope

> All API  URL's are constructed using one of the following patters:

Reviewer: Gal reikėtu susieti URL kelio elementus su [xsd:token](https://www.w3.org/TR/xmlschema-2/#token)? Kaip  tai yra padaryta su [prof:hasToken](https://www.w3.org/TR/dx-prof/#Property:hasToken)?

Arba, gal būti galima remtis, kaip XML specifikacijoje yra apibrėžtas Name:

https://www.w3.org/TR/2000/WD-xml-2e-20000814#dt-name

Taip pat siūlyčiau pateikti nuorodą į DSA specifikaciją, kurioje yra paaiškinti kodiniai pavadinimai.

> All API  URL's are constructed using one of the following patters:

Reviewer: Siūlyčiau URL Path struktūrą apibrėžti keliais gabalais, kad nereikėtu pateikti tokių ilgų nuorodų. Pavyzdžiui:

Galima pradžiai apibrėžti apllinką, duomenų paslaugą ir veiksmą:

    https://{env}/{service}/{action}

Toliau galima detalizuoti kiekvieną atskirai:

Aplinka (`{env}`): Siūlyčiau pateikti nuorodą į "Concepts / Environment" ir ten pateikti detalesnį aplinkų aprašymą.

Duomenų paslauga (`{service}`):

    {group}/{form}/{publisher}/{is}/{service}/{version}

Vėl, kiekvienas elementas gali būti aprašytas atskirame "Concepts / {elem}" skyrelyje.

Toks aprašymas atitiktu specifikacijoms būdingą [EBNF](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form) (ISO/IEC 14977:1996) principą. Tačiau siūlyčiau išlaikyti [OAS Path Templating](https://spec.openapis.org/oas/latest.html#path-templating) sintaksę.

- Get list of objects:

>   <span class="green">domain</span> `/`

Reviewer: Gal geriau reikėtu vadinti "environment", nes "domain" kertasi su veiklos sritimi. Jei turima mintyje DNS Domain, tada per plati sąvoka, nes UDTS atveju, eina kalba tik apie aplinkas (TEST, PROD) ir saugumo sritis (vidiniai, išoriniai vartai).

Siūlau aplinkas iškelti į atskirą skyrių "Concepts / Environment", kur reikėtu apibrėžti aplinkų matricas, siejant su prieigos teisėmis:

| Saugumo sritis \\ Aplinka | TEST | PROD |
| ------------------------- | ---- | ---- |
| protected                 | test-apigw.gov.lt | apigw.gov.lt |
| public                    | test-api.gov.lt | api.gov.lt |
| open                      | get-test.data.gov.lt | get.data.gov.lt |

Kur Saugumo sritis atitinka DSA specifikacijoje apibrėžtus prieigos lygmenis:

https://ivpk.github.io/dsa/draft/prieiga.html

Analogiškai, aplinkas reikėtu apibrėžti ir autorizacijos dalyje, nurodant skirtingus autorizacijos serverių adresus, pagal saugumo sritis / aplinkas.

>   `datasets/`

Reviewer: Ši dalis taip pat kintanti, gali būti:

- datasets - UDTS duomenų paslaugos
- services - ne UDTS, legacy duomenų paslaugos

  <span class="green">form</span> `/`

>   <span class="green">org</span> `/`

Reviewer: ```suggestion
  <span class="green">publisher</span> `/`
```

Siūlau nurodyti tikslesnį pavadinimą, kuris atitiktų DCAT-AP-LT specifikaciją.

>   <span class="green">catalog</span> `/`

Reviewer: ```suggestion
  <span class="green">{is}</span> `/`
```

Per UDTS publikuojamos tik informacinių sistemų paslaugos, tačiau nėra publikuojamos visų katalogų paslaugos.

Taip pat gali atsirasti posistemė `{is}/{subsys}`.

  `datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model</span> `/`
  `properties/`
  <span class="green">property</span>

  <span class="green">domain</span> `/`
  `/datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model</span> `/`
  <span class="green">id</span> `/`
  `properties/`
  <span class="green">property</span>

- Perform a custom action on a collection or on a single object:

  <span class="green">domain</span> `/`
  `/datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model</span> `/`

>   `actions/`

Reviewer: Tai yra MAJOR pokytis, nes veiksmai buvo nurodomi `/:{action}/{argument}` forma. Ar tikrai yra noras daryti MAJOR pokyčius specifikacijoje?

  <span class="green">domain</span> `/`
  `/datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model</span> `/`
  <span class="green">id</span> `/`
  `actions/`
  <span class="green">action</span>

- Standar query parameters MUST be used.

  <span class="green">domain</span> `/`
  `datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model<span> `?`
  <span class="green">query<span>
  
   <span class="green">domain</span> `/`
  `datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model<span> `/`
  <span class="green">id</span> `?`

>   <span class="green">query<span>

Reviewer: Ar tikrai `query` galima naudoti objekto lygmeniu? Kokiais atvejais tai galėtu būti naudojama?

Author: Kai atsakymo duomenys filtruojami ir ne visi laukai iš modelio yra grąžinami atsakyme

Reviewer: Tada reikėtu pateikti query parametrus ir prie https://ivpk.github.io/uapi/draft/#tag/object/operation/getOne, dabar ten nėra nurodyta, kad galima naudoti Query params.

Taip pat reikėtu patikslinti, kaip turėtu veiki API tokiais atvejais:

- Jei filtras grąžina daugiau nei vieną objektą, tai turėtu grąžinti tik tą objektą, kurio `id` nurodytas Path dalyje?
- Jei filtras grąžina duomenis, tarp kurio nėra `id` Path dalyje, nors be filtro toks objektas yra. Ar turėtu grąžinti 404 not found ar kažką kita?

Author: Reikia apra6omojoje dalyje pašalinti querie iš vieno objekto URL.

  
- Retrieve object or property based on scope:

  <span class="green">domain</span> `/`
  `datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model<span> `/`

>   `scopes/@`

Reviewer: Dar vienas nesuderintas pokytis, iki šiol buvo numatyta, kad custom scope filtrai nurodomi `/@{scope}` forma. Ar tikrai keičiame į `/scopes/@{scope}`?

  <span class="green">scope<span>
  
   <span class="green">domain</span> `/`
  `datasets/`
  <span class="green">form</span> `/`
  <span class="green">org</span> `/`
  <span class="green">catalog</span> `/`
  <span class="green">service</span> `/`
  <span class="green">version</span> `/`
  <span class="green">dataset</span> `/`
  <span class="green">model<span> `/`
  <span class="green">id</span> `/`
  `scopes/@`
  <span class="green">scope<span>  

- The following pattern MAY be used for APIs, mostly legacy APIs, that do not support the standard UAPI exchange protocol:

  <span class="green">domain</span> `/`
  `/services/`
  <span class="green">form</span> `/`

>   <span class="green">org</span> `/`

Reviewer: ```suggestion
  <span class="green">publisher</span> `/`
```

>   <span class="green">catalog</span> `/`

Reviewer: ```suggestion
  <span class="green">is</span> `/`
```

Su neprivalomu `/{subsys}`.

  <span class="green">service</span> `/`
  <span class="green">version</span> `/`

This pattern is a valid alternative in the current version of the specification, but it SHOULD be considered deprecated in future versions.

- Ownership and naming

> SDSA has the exclusive right to register `org` and `catalog` values in the data.gov.lt catalog and to resolve disputes among Agents.

Reviewer: Reikėtu apibrėžti, kas yra "SDSA"? Ir klausimas ar kaip jis susijęs su UDTS? Planuojama ŠDSA dalį iškelti į Spintos dokumentaciją.

Author: Patikslinta čia State Digital Solution Agency

> SDSA has the exclusive right to register `org` and `catalog` values in the data.gov.lt catalog and to resolve disputes among Agents.

Reviewer: ```suggestion
SDSA has the exclusive right to register `publisher` and `is` values in the data.gov.lt catalog and to resolve disputes among Public Organizations.
```

Siūlau naudoti tikslesnes sąvokas, atitinkančias DCAT-AP-LT.

An Agent has the right to choose `service`, `dataset`, `model`, and `property` values, provided that these values conform to the syntax rules of this and [DSA specification](https://ivpk.github.io/dsa/).

Meaning of dynamic URL parts:

> - <span class="green">domain</span> - domain of the data provision platform. Depending on selected data access mode, either external or internal gateway domain address. 

Reviewer: Ar nereikėtu išvardinti tiksliau:

- `open` - atvirų duomenų Saugykla (get.data.gov.lt)
- `public` - išoriniai vartai (api.gov.lt)
- `protected` - vidiniai vartai (apigw.gov.lt)
- `private` - konkrečios informacinės sistemos domenas

Kažkur matyt reikėtu paminėti ir TEST aplinkas:

- `open` - atvirų duomenų Saugykla (get-test.data.gov.lt)
- `public` - išoriniai vartai (test-api.gov.lt)
- `protected` - vidiniai vartai (test-apigw.gov.lt)
- `private` - konkrečios informacinės sistemos domenas

Ar bent jau pateikti nuorodą į DSA:

https://ivpk.github.io/dsa/draft/prieiga.html

- <span class="green">form</span> - The legal form of an organization. This is a controlled vocabulary with only two permitted values:
	`gov` — government sector

> 	`com` — private or other non-government sector

Reviewer: Čia tikriausiai turėtu būti tik registruoti juridiniai asmenys, kurie nėra valstybinio sektoriaus.

The `gov` value is understood as being under the regulation of the Law on the Management of State Information Resources of the Republic of Lithuania.

> The `com` value covers all other cases.

Reviewer: Turėtu būti dar ir `org`, kuris padengia visus kitus atvejus, tokius kaip nepelno siekiančias organizacijas ir pan.

`com` yra JAR objektas. `org` visa kita.

> - <span class="green">org</span> - A short code name of an organization or agent. It represents the class [foaf:Agent](https://xmlns.com/foaf/spec/#term_Agent) ([dc:publisher](http://purl.org/dc/elements/1.1/publisher)) and its property [skos:notation](https://www.w3.org/TR/skos-reference/#notations). 

Reviewer: `foaf:Agent` yra labai platus, galbūt reikėtu patikslinti, kad `gov` atveju, kalba eina tik apie `cv:PublicOrganisation`, jei `{form}` yra `gov`?

- <span class="green">catalog</span> - Catalog is a short code identifier representing a catalog, sub-catalog, information system,
 or sub-information system that provides data, corresponding to the class [dcat:Catalog](https://www.w3.org/TR/vocab-dcat-3/#Class:Catalog) and its property [skos:notation](https://www.w3.org/TR/skos-reference/#notations),

>  and in the context of DCAT-AP-LT v2 most often representing the [dcataplt:InformationSystem](https://ivpk.github.io/DCAT-AP-LT/en/##IS) class.

Reviewer: `Catalog` ir `InformationSystem` yra labai skirtingi dalykai. Visos `InformationSystem` yra `Catalog`, bet ne visi `Catalog` yra `InformationSystem`.

Todėl per UDTS būtent publikuojam tik `InformationSystem`. Todėl siūlyčiau čia ir kitose vietose nurodyti `{is}` ir `dcataplt:InformationSystem`.

<span class="green">catalog</span> values used by existing data catalogs:

`adp` — Lithuanian Open Data Catalog

`ror` — Lithuanian Protected Data Catalog

> `isris` — Lithuanian Protected Data Catalog

Reviewer: Per UDTS nėra publikuojami visi katalogai, yra publikuojamos tik informacinės sistemos registruotos ISRIS kataloge.

Kiek žinau nėra numatyta galimybė, publikuoti UDTS paslaugas per kitą, nei ISRIS katalogą. Todėl šiuo atveju, manau turėtu būti nurodoma tik informacinė sistema, kuri yra registruota ISRIS kataloge ir kuriai suteiktas kodinis pavadinimas, be galimybės publikuoti UDTS paslaugas iš kitų katalogų.

Example:

```text

> .../datasets/gov/rc/jadis/uapi/1/at280/DalyvioAsmensIsrasas

Reviewer: ```suggestion
/datasets/gov/rc/jadis/at249/1/israsas/DalyvioAsmensIsrasas
```

Siūlau nenaudoti `uapi`, kaip Duomenų paslaugos kodinio pavadinimo, nes visos `/datasets/*` paslaugos yra UDTS. Geriau naudoti kodinį pavadinimą, kuris apibūdina Duomenų paslaugą.

Gal būtų geriau naudoti atvirų duomenų saugyklos pavyzdžius, kuri taip pat įgyvendina UDTS? Tokiu būdų bus galima gyvai išsibandyti, nesudarant jokių sutarčių?

```

In this example:

> `datasets` - defines it as dataset for a data retrieval via UDTS

Reviewer: Defines what?

URL Path parametras yra apibrėžtas kaip `group`:

https://github.com/ivpk/uapi/blob/bc47d2b1740928856924c7ba7b06e66c2485b7ac/draft/openapi/components/parameters/group.yaml#L1

Konkrečiai `datasets` group yra rezervuota UDTS duomenų paslaugos, publikuojamoms automatizuotu būdu iš Katalogo.

`gov` - defines it as government sector data

> `rc` - is the code name of organisation 

Reviewer: Code name of a publisher.

`jadis` - is understood as an information system,

`uapi` - is the data service name.

`1` - is the version of the service

`at280` - is the name of the dataset

`DalyvioAsmensIsrasas` - is the name of the model

> - <span class="green">service</span> - A short code name of a data service. It represents the class dcat:DataService and its property [skos:notation](https://www.w3.org/TR/skos-reference/#notations).

Reviewer: Kodėl `dcat:DataService` yra be nuorodos, o `skos:notation` yra su nuroda? Gal reikėtu pateikti nuorodas į taikymo profilį:

https://ivpk.github.io/DCAT-AP-LT/##DuomenuPaslauga

Nurodant, kad `service` yra `dcat:endpointURL`? Tiksliau, `dcat:endpointURL` kelio dalis?

Gal būt reikėtu pateikti ryšį tu EIRA "Digital Solution" Application Service, prieinamas per "Machine to Machine Interface" Application Interface:

<img width="848" height="624" alt="Image" src="https://github.com/user-attachments/assets/7760c78c-cd5c-4e40-81f3-1fc20a882be4" />

<https://cartool-ec.github.io/EIRA/index.html?view=id-6a8b3a5a>

Kas taip pat yra aprašyta DTFS:

> 14. Kuriant arba modernizuojant Išteklius, rekomenduojama sprendimo architektūrą grįsti „pirmiausia API“ (angl. API-first) principu, kuris apibrėžia du sluoksnius: aplikacijų programavimo sąsajos (angl. backend) bei naudotojo sąsajos (naršyklės kliento, programėlių ar kt.) (angl. frontend). Rekomenduojama naudoti UDTS kaip pagrindinę Ištekliaus aplikacijų programavimo sąsają, o naudotojo sąsają laikyti su kitais Ištekliais lygiaverčiu duomenų gavėju.

<https://e-seimas.lrs.lt/portal/legalAct/lt/TAD/TAIS.445504/asr>

Tačiau, plačiau apie tai siūlyčiau aprašyti prie "Concepts / Service" skyriaus, o čia pateikti tik nuorodą.

- <span class="green">version</span> - The data schema version of a data service. It represents the MAJOR version number according to the 
[Semantic versioning v2.0.0](https://semver.org/) version management rules.

The following rules apply:

> a.  If a [dcat:Dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) receives a new MAJOR version, it does not affect the dcat:DataService URI; it only affects the dcat:downloadURL of the dcat:DataService.

Reviewer: Gal būt reikėtu nuorodas pateikti į DCAT-AP-LT? Tik problema, kad DCAT-AP-LT nepalaiko tiesioginių nuorodų į property elementus.

Rekomenduoju sieti ne su DCAT, o su OAS. Semantinę prasmę nusako DSA.

Konkrečiai versijos elementas atitinka OAS `Info.version` atributą:

https://spec.openapis.org/oas/latest.html#info-object

b. If a dcat:Dataset does not receive a new MAJOR version, but the [dcat:DataService](https://www.w3.org/TR/vocab-dcat-3/#Class:Data_Service) 
that [dcat:servesDataset](https://www.w3.org/TR/vocab-dcat-3/#Property:data_service_serves_dataset) that [dcat:Dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) receives 

> a MAJOR update of the OAPI specification, then the MAJOR version number used in the URI version segment MUST be updated accordingly.

Reviewer: ```suggestion
a MAJOR update of the UAPI specification, then the MAJOR version number used in the URI version segment MUST be updated accordingly.
```

> - <span class="green">dataset</span> - A short code name of a specific dataset from a catalog. It MAY also represent a sub-dataset. 

Reviewer: Kaip ir minėjau anksčiau, tokius atvejus galima apibrėžti naudojant EBNF principą, kuri `{dataset}` gali būti skaidomas į: `{dataset}/{sub-dataset}`.

Author: Bet ar tada neatsiranda pas mus ambiguous path kai ten gali būti ir `sub-dataset` ir `model` ir jie abu yra `string`. Aš suprantu, kad teoriškai model turi prasidėti didžiąja reide, bet linter ir kitiems validatoriams nepatinka

> It represents the class [dcat:Dataset](https://www.w3.org/TR/vocab-dcat-3/#Class:Dataset) and its property [skos:notation](https://www.w3.org/TR/skos-reference/#notations).

Reviewer: Rekomenduoju sieti ne su DCAT, o su OAS. Semantinę prasmę nusako DSA.

Konkrečiai Dataset atveju, OAS atitikmuo būtų Tag elementas:

https://spec.openapis.org/oas/latest.html#tag-object

Ir tikriausiai galima nurodyti `kind: dataset`.

Siūlyčiau įtraukti naują "Concepts / Dataset" skyrių, kuriame būtų plačiau paaiškinta, kas yra "Dataset" UDTS kontekste ir kas yra "Virtual Dataset", kaip  tai siejasi ir kaip tai aprašoma OpenAPI pagalba, naudojant Tag objektus.

Gal būt būtų galima susieti su [DDD Aggregate](https://martinfowler.com/bliki/DDD_Aggregate.html)? Kadangi EIF rekomenduoja remtis DDD.

Taip pat siūlyčiau pateikti sąsają su EIRA "Dataset" Data Object:

<img width="553" height="419" alt="Image" src="https://github.com/user-attachments/assets/94f61abe-d2d4-4d83-97a3-b72dfcc935af" />

<https://cartool-ec.github.io/EIRA/id-3d54d4b1/views/id-13cfc7d1.html>

Paaiškinant, kuo skiriasi Dataset nuo Virtual Dataset, kad pagrindinis tikslas teikti "Master Data", kurie priklauso "Base Registry", o Base Registry atitinka IS?

Author: Atnaujinti aprašymą ir paaiškinimus susiejant su `OAS:tag` ir atnaujinti šabloną su hierarchical tag'ais kur top level yra `dataset` ir sub yra objects/object/resource. Atskirus URI elementus išskaidyti į Concept elementus. Taip pat aprašyti modelius kaip dataset - yaml struktūros nekeisti, tik aprašomojoje dalyje paaiškinti, taip paaiškinti kad ne visi modeliai yra šitos paslaugos dataset, ir gali būti negrąžinami per path. Modelis bus `component/schema`. Modeliai ir jų prop turėtų būti pateikiami inline nenaudojant $ref, o susiejimą pateikiant per `_type`

- <span class="green">model</span> - A SHACL validation shape ([shacl:NodeShape](https://www.w3.org/TR/shacl/#node-shapes)) that defines constraints for a set 

> of objects conforming to the same data model, with [shacl:targetClass](https://www.w3.org/TR/shacl/#targetClass) pointing to the corresponding [rdfs:Class](https://www.w3.org/TR/rdf-schema/#ch_class).

Reviewer: Rekomenduoju sieti ne su DCAT, o su OAS. Semantinę prasmę nusako DSA.

Konkrečiai Model atveju, OAS atitikmuo būtų Schema objektas pateikiamas `components.schemas`. Kad būtų galimybė atskirti modelius nuo kitų schemų, `components.schemas` modeliai turi prasidėti didžiąja raide, pagal Model kodinių pavadinimų reikalavimus, o visos kitos schemos turėtu būti pateikiamos arba inline būdu arba taikant property kodinių pavadinimų taisykles.

Taip pat gal būt galima sieti su EIRA "Data Model" Data object?

<img width="194" height="370" alt="Image" src="https://github.com/user-attachments/assets/2c9978c7-e61c-48d0-9360-27426601fc37" />

<https://cartool-ec.github.io/EIRA/index.html?view=id-13cfc7d1>

Kur "Data Model" apibrėžimas skamba taip:

> Data Model ABB is a Data Object aiming to be a collection of entities, their properties and the relationships among them, which focus at formally representing a domain, a concept or a real-world thing.

UDTS kontekste, Modelis atitinka TOGAF "Logical Data Component":

> A data structure composed of logically-related data entities.

<https://pubs.opengroup.org/togaf-standard/architecture-content/chap02.html#chap02>

Kas iš esmės yra Aggregate "Data entity", kuris gali būti sudarytas iš kelių skirtingų duomenų esybių susijusių ryšiais.

> Data Entity - Represents data that is recognized by the business as a distinct concept.

<https://pubs.opengroup.org/togaf-standard/architecture-content/chap02.html#chap02>

            | _id        | string  |          | /id,code
            | _revision  | string  |          | /id,code/major,minor
            | id         | integer |          |
            | code       | string  |          |
            | major      | integer |          |
            | minor      | integer |          |
            |            |         |          |
Str         |            |         | name     |
            | _id        | string  |          | /=name
            | _revision  | string  |          | /=name/=version
            | name       | string  |          |
            | version    | string  |          |
            |            |         |          |
Base32      |            |         | code     |
            | _id        | base32  |          | /=BASE32(code)
            | _revision  | string  |          | /=name/=version
            | code       | string  |          |
            |            |         |          |
CompBase32  |            |         | id,code  |
            | _id        | base32  |          | /=BASE32(CBOR(id,code))
            | id         | string  |          |
            | code       | string  |          |

>             |            |         |          |

Reviewer: Turėtu būti pateikta nuoroda į "Identifier" skyrių, kuriame plačiau paaiškinta, apie identifikatoriaus sandarą.

- <span class="green">revision</span> - A unique identifier of a historical object version. 

> Revision semantics MUST follow the Linked Data Event Streams ([LDES](https://semiceu.github.io/LinkedDataEventStreams/)) model. UUID format is RECOMMENDED.

Reviewer: LDES neapibrėžia konkrečių versijavimo atributų, tik leidžia nurodyti, kokie atributai yra naudojami. Pačius atributus apibrėžia DCAT:

https://www.w3.org/TR/vocab-dcat-3/#version-info

Ir VSSA Core Vocabulary:

https://ivpk.github.io/core_vocabularies/vssa/risr/core/1.0/index.html

Kur yra apibrėžtas `_revision`.

- <span class="green">property</span> - A subresource used to retrieve the value of a property, 
especially where the property value is a large object, linked file, or array. It represents an RDF property, 

> such as [owl:ObjectProperty](https://www.w3.org/TR/2004/REC-owl-semantics-20040210/#owl_ObjectProperty) or [owl:DatatypeProperty](https://www.w3.org/TR/2004/REC-owl-semantics-20040210/#owl_DatatypeProperty).

Reviewer: Siūlyčiau pateikti nuorodą į "Subresource", kur yra plačiau paaiškinta, kas tai yra.

Example:

> `<owl:ObjectProperty rdf:about="#hasBankAccount"/>`

Reviewer: Nesuprantu, kodėl pateiktas RDF/XML pavyzdys, kai UDTS naudoja JSON-LD?

> Example URI:

Reviewer: Pavyzdyje žemiau pateiktas ne URI, o Distribution  dowloadURL.

```

> https://api.gov.lt/datasets/com/cemety/ldis/1/kapines/vilnius/Cemeteries/123/properties/hasLabel

Reviewer: Pavyzdys neatitinka `property` kodinio pavadinimo reikalavimų. Taip pat siūlyčiau, nepateikti konkrečios aplinkos, nes jų yra kelios. Gali klaidingi, kad yra naudojama tik `api.gov.lt`, nes vidiniuose vartuose yra `apigw.gov.lt`.

```suggestion
/datasets/com/cemety/ldis/1/kapines/vilnius/Cemeteries/123/has_label
```

```

> - <span class="green">action</span> - An operation performed on a collection of objects or on a single object. It MUST start with `:`.

Reviewer: Siūlau pateikti nuorodą į "Action" skyrių, kur paaiškinta plačiau kas tai yra, o čia nebūtina detalizuoti.

> - <span class="green">scope</span> - Access based filter using OAuth 2.0 scopes. It MUST start with `@`.

Reviewer: Reikėtu pateikti nuorodą į "Authorization / Scope", kur plačiau paaiškinta, kas tai yra. O paaiškinant reikia nurodyti, kad tai yra custom  scope, kuris filtruoja duomenis ir leidžia enforcinti filtrą per access token.

> Examples of API endpoints:

Reviewer: Nesuprantu apie ką ši dalis?

Pavyzdžiuose naudojamos naujos sąvokos tokios kaip `concept`, nors panašu, kad kalbama apie Model?

Taip pat pavadinta "API endpoints", nors DCAT nurodo, kad API endpint yra Data Service root URL. Tai kaip suprantu, čia kalba eina labai apie Actions (kurie gali būti implict arba explicit).

Nelabai aišku, kodėl reikia tai kartoti, nes analogiški pavyzdžiai jau yra pateikti aukščiau?

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>
`/actions/`
<span class="green">action</span>  - custom action

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>
`/scopes/@`
<span class="green">scope</span>  - custom filter

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>`/`
<span class="green">reference</span>  - getone

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>`/`
<span class="green">reference</span>
`/properties/`
<span class="green">property</span>  - getall on property, for file and array types

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>`/`
<span class="green">reference</span>`/:changes`  - changes on single object

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>`/`
<span class="green">reference</span>
`/actions/`
<span class="green">action</span>  - custom action on single object

`/`<span class="green">dataset</span>`/`
<span class="green">concept</span>`/`
<span class="green">reference</span>
`/revision/`
<span class="green">revision</span>  - specific version of an object
  

> Static parts:

Reviewer: Ar neturėtu būti išlaikomas eiliškumas? URL elemntų sąraše, UDTS Duomenų paslaugų grupė eina iš karto po aplinkos. Todėl, gal turėtu būti išlaikytas eiliškumas?

Be to, siūlyčiau, kaip ir kitose vietose, pateikti nuorodą į atskirą "Concepts / Group" skyrių?

- `datasets` - dataservice enpoints used for data retreival via UAPI service.
- `services` - services endpoint, that does not follow UAPI requirements, used for existing legacy API services.
- `properties` - endpoint used to retrieve properties of an object.
- `actions` - endpoint that allows to perform custom actions on a collection or on a single object
- `revision` - endpoint that allows to retrieve a different version of the element

> - `scopes` - endpoint that allows to define data filtering based on access

Reviewer: Siūlau neįvedinėti naujų sintaksės elementų. Išvardinti dalykai:

- `properties`
- `actions`
- `revision`
- `scope`

Turi implicit formas. Siūlau jas ir palikti, kaip buvo, naujų sintaksės elementų neįvedinėti.

The following URL components are mandatory for all datasets and services endpoints:

  <span class="green">form</span>
  
  <span class="green">org</span>
  
  <span class="green">catalog</span>
  
  <span class="green">service</span>

All other components are optional and depend on the level of access being addressed, such as 

  <span class="green">dataset</span>

  <span class="green">model<span>
  
  <span class="green">revision</span>
  
  <span class="green">property<span>
  

>   <span class="green">action</span>

Reviewer: Gal būt nevadinčiau to "optional" komponentais. Gal būt geriau įvardinti, kad UDTS URL kelias yra sudarytas iš dviejų dalių:

- Endpoint URL - Data Service root URL.

- Action - dinaminė dalis, gali būti vykdomu įvairūs veiksmai, kurie kelio dalyje pateikiami implicit arba explicit forma?

Gal būt reikėtu pateikti nuorodą į "Concepts / Action", kur būtų plačiau paaiškinta apie veiksmus?

> API URL example:

Reviewer: Tai nėra URI, o Download URL.

```
https://get.data.gov.lt/datasets/gov/rc/ar/uapi/text_with_coordinates/Address/e96cc0cc-08be-460d-a887-98f80612a402
```

When interpreting URL, there are few distinctions in each component, that might
help separate components presented in URL:

- Namespace components <span class="green">form</span>,
  <span class="green">org</span>, <span class="green">catalog</span>,
  <span class="green">service</span>  and <span class="green">dataset</span> must 
  start with lowercase ascii letter and must always follow one after
  the other, with catalog exception, where
  dataset might be used without catalog.

- <span class="green">model</span> name must always start with uppercase ascii
  letter and can be used to separate namespace components from model.

- <span class="green">property</span> must always start with lowercase ascii
  letter and cannot contain `-` character to separate property from id.

> - <span class="green">action</span> has an explicit `/actions/` endpoint.

Reviewer: Siūlyčiau nuo to ir pradėti. Tik gal būt iškelti tokį paaiškinimą į "Architecture" dalį, pateikiant nuorodas į "Concepts", kur paaiškinama kiekviena sąvoka detaliau.

Siūlau pirmiausiai skaidyti į dvi dalis:

- Endpoint URL - Duomenų paslaugos šakninė dalis, įskaitant versiją.
- Action - viskas, kas eina po Endpoint URL.

Tada galima skaidyti šias dalis dar smulkiau.

> ## Semantic interpretation of API URLs

Reviewer: Šis skyrius kartojasi, nes jau buvo nurodyta semantinė prasmė, prie kiekvieno URL kelio elemento. Nežinau, ar būtina kartoti?

The API URL structure defines an access path and governance namespace. It does not, by itself, define an RDF graph structure.

The semantic meaning of URL segments is expressed through [JSON-LD Context](https://www.w3.org/TR/json-ld/#the-context) and [RDF](https://www.w3.org/RDF/) mappings. In particular:

`org` identifies a foaf:Agent

`catalog` identifies a dcat:Catalog

`service` identifies a dcat:DataService

`dataset` identifies a dcat:Dataset

`model` identifies an owl:Class

`property` identifies an RDF property.

> ## URI and Semantic Resource Identification Overview

Reviewer: Siūlyčiau iškelti visą šią dalį prie "Concepts / Identifier"?

This specification distinguishes between:

**vocabulary resources**, such as `classes`, `properties`, and `concepts`;

and

> **entity resources**, such as identified instances of `datasets`, `catalog` entries, and `domain` objects.

Reviewer: Jei teisingai suprantu, tai norima išskirti URL ir URI.

URL- yra Resource Locator - nurodo ištekliaus fizinę vieną, siejama su Data Service endpoint URL ir Distribution download URL.
URI - yra Resource Identifier - nurodo objekto globalų identifikatorių, neprivalomai, tačiau gali būti siejamas ir su fizine objekto reprezentacija, nukreipiant į URL.

Taip pat reikėtu atskirti URI/URL kurie yra skirti metaduomenims (klasės, savybės, sąvokos) ir URI/URL kurie yra skirti duomenims (esybės, objektai, reikšmės).

UDTS yra skirtas duomenims, o ne metaduomenims.

Todėl, siūlyčiau lyginti kuo skiriasi Locator ir Identifier, o ne Vocabulary ir Entity.

These resource types MUST be identified using consistent URI patterns.

- Vocabulary resources

> **Vocabulary resources** define the semantic layer of the system, including ontologies and controlled vocabularies.

Reviewer: Taip, tačiau šiuo atveju, kalba eina ne apie metaduomenis, o apie duomenis. Todėl nematau, kad verta pateikti "Vocabulary resources" atvejį? Per UDTS bus pateikiami tik "Entity resources"?

Vocabulary resources SHOULD use the following pattern:

  `https://data.gov.lt/id/`
  <span class="green">vocab</span>`/`
  <span class="green">termType</span>`/`
  <span class="green">termName</span>

Where:

- <span class="green">vocab</span> — a code name of the data domain; it MAY align with EU or W3C vocabulary names and corresponds to prof:hasToken

- <span class="green">termType</span> — type of vocabulary term; MUST be one of:

  - `class`

  - `property`

  - `concept`

- <span class="green">termName</span> — local name of the term

Examples
https://data.gov.lt/id/jar/class/JuridinisAsmuo
https://data.gov.lt/id/jar/property/kodas
https://data.gov.lt/id/evrk/concept/6201

- Entity resources

**Entity resources** represent identifiable instances of real-world or digital objects.

Entity resources MUST use the following identifier pattern:

  `https://data.gov.lt/id/`
  <span class="green">vocab</span>`/`
  <span class="green">entityType</span>`/`

>   <span class="green">id</span>

Reviewer: Gal reikėtu remtis 10 rules for persistent URIs:

<img width="2943" height="1113" alt="Image" src="https://github.com/user-attachments/assets/da17d47f-e071-405b-b92a-947ccc79f729" />

-- https://interoperable-europe.ec.europa.eu/sites/default/files/document/2013-02/D7.1.3%20-%20Study%20on%20persistent%20URIs.pdf

Ir nurodyti tokį pavyzdį:

    http://{domain}/{type}/{concept}/{reference}

Nurodant, kad UDTS atveju, naudojama `type=id` dalis, neapimant metaduomenų (Vocabulary resources)?

Taip pat reikėtu nurodyti, kad URI vardų erdvė yra apibrėžiama KDSA, kuris yra pateikiamas prie IS?

Teisingas pavyzdys:

    https://data.gov.lt/id/{concept}/{reference}/{revision}

Where:

- <span class="green">vocab</span> — a code name of the data domain

- <span class="green">entityType</span> — a semantic identifier of the entity type (typically aligned with an owl:Class)

- <span class="green">id</span> — a globally unique and persistent identifier of the entity

Versioned entity resources

If versioning is required, entity resources MAY include a revision identifier:

https://data.gov.lt/id/{vocab}/{entityType}/{id}/{revision}

  `https://data.gov.lt/id/`
  <span class="green">vocab</span>`/`
  <span class="green">entityType</span>`/`
  <span class="green">id</span>`/`

>   <span class="green">revision</span>

Reviewer: Turėtu būti:

    https://data.gov.lt/id/{concept}/{reference}/{revision}

Where:

- <span class="green">revision</span> — an identifier of a historical version of the entity

> Revision semantics MUST follow the Linked Data Event Streams ([LDES](https://semiceu.github.io/LinkedDataEventStreams/)) model.

Reviewer: LDES neapibrėžia konkretaus versijavimo principo, tik leidžia nurodyti, kaip versijavimas yra įgyvendintas. Čia gal reikėtu remtis DCAT:

https://www.w3.org/TR/vocab-dcat-3/#version-info

Pateikiant nuorodą į "Concepts / Revision".

Relationship with API URLs

API URLs define an access path and governance namespace, while URIs define persistent semantic identifiers.

> An API URL MAY resolve to a representation of an entity resource identified by a URI.

Reviewer: API URL MUST resolve to a representation,  but URI might de-reference to such representation.

For example:

> ```code

Reviewer: ```suggestion
```http
```

## Revision

All objects must have a revision number provided as reserved `_revision`
property name.

Revision is a string, that identifies object version. This might be a UUID
string.

When update action is made on an object, revision number must always be
provided, in order to compare object version received by the Client and object
version in the Resource. This check is needed to avoid conflicts, when two or
more concurrent requests try to update same object.

When object is changed, a new revision number must also be generated.

Revision number is also used, references older object version if Resource
stores older object versions:

```uri
/datasets/gov/rc/ar/uapi/text_with_coordinates/Country/e96cc0cc-08be-460d-a887-98f80612a402/revision/dd22f1b4-09c2-48ee-bf7a-7bf082da9940
```

Second UUID number is a revision ant following request retrieves an older
object version.

Revision number is retrieved as part of the `changes` request for the object.

Revision number is also used as [HTTP ETag
header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) to
indicate if resource is outdated.

> ## Maturity

Reviewer: Reikėtu pateikti pavyzdžius, ne tik `Model`, bet ir `ref` atveju.

> ## Maturity

Reviewer: Taip pat siūlyčiau pagalvoti, kaip išreikšti DSA:level OpenAPI sintakse. OpenAPI leidžia daryti išplėtimus, todėl siūlyčiau apibrėžti, kaip brandos lygis nurodomas OAS formatu, pavyzdžiui:

```json
{
  "components": {
    "schemas": {
      "City": {
        "type": "object",
        "x-level": 4,
        "properties": {
          "country": {
            "type": "object",
            "x-level": 3
          }
        }
      }
    }
  }
}
```

Tokiu būdu galima bus atlikti OAS <-> DSA konversiją, neprarandant metaduomenų.

> Metadata and data maturity is defined in DSA<>. To comply to UDTS specification maturity level MUST be at least 2. 

Reviewer: Ar mes galime nustatyti reikalavimus? Mano supratimu, brandos lygis tiesiog nurodo faktą. Jei duomenų nėra, tai juos vis dar galima teikti per UDTS, net ir mažesniu brandos lygiu?

Kadangi referuojama į DSA, gal reikėtu pateikti nuorodą į brandos lygius?

https://ivpk.github.io/dsa/1.1/branda.html

Author: Atnaujinta nuoroda į DSA

Highest maturity level be should be strived for.

Data values might have different forms depending on data maturity level. There

> are five maturity levels:

Reviewer: Čia gal būt kalbama ne apie visus data values, o tik apie identifikatorių reikšmes "identifier values"?

- `0` - data is not available, even if property is present, in such cases all

>   values are set to `null`.

Reviewer: Gal reikėtu patikslinti, kad nulinio brandos lygio atveju, gali nebūti jokių reikšmių, t.y. gali nebūti jokio lauko, kuris atitiktu objekto identifikatorių.

Jei vis tik laukas yra pateiktas ir neturi jokių reikšmių, tik tada nulinis brandos lygis nurodytu, kad visos reikšmės yra `null`.

- `1` - data is not machine readable, usually this means, that there is no
  clear structure or consistency in data values. For example a number might

>   have different units, bet unit itself is not known. Or date and time values

Reviewer: Kadangi čia kalbame apie identifikatorius, kurie neturi vienetų, gal reikėtu nurodyti L103 ir L104 atvejus:

- L103 - neįmanomas jungimas, `ref` laukų atveju.
- L104 - identifikatorius nėra unikalus,  `model` atveju.

Visais kitais atvejais, UDTS data payload struktūros nekeičia, lyg ir?

  might be given using different formats.

- `2` - data is consistent ant machine readable, but values are given in a
  non-standard way. For example integer values, might be given as strings, o

>   date values might be given in a non-standard format.

Reviewer: Gal būt vėl reikėtu kalbėti apie pirminius (`model`) ir išorinius (`ref`) laukus ir jų reikšmes, remiantis DSA:

- L204 - nepatikimi identifikatoriai, tarkime naudojamas kintantis pavadinimas, vietoj nekeičiamo identifikatoriaus.
- L206, L209 - nenurodytas susiejimas, kai nėra nurodyta, su kokiu ref ar base modeliu yra siejamas identifikatorius.
- L207 - neatitinka modelio bazės, kai identifikatorius nėra išduotas bazinio registro.

> - `3` - objects do not have a global object identifier in UUID format, this

Reviewer: Objekto identifikatorius nebūtinai turi būti UUID formatu.

  usually applies to categorical data, references and `_id` property of an
  object. In such cases it is not possible to retrieve a single object by its

>   global identifier, `_id` might not be given.

Reviewer: Šiuo atveju reikia žiūrėti į DSA L301 Nėra globalaus identifikatoriaus.

Taip pat reikėtu patikslinti, kad data payload neturi `_id` ir UDTS užklausos su vienu objektu, privalo naudoti URL Query filtrus. Draudžiamos užklausos su vienu objektu, kur identifikatorius nurodomas per URL path.

- `4` - objects have global identifiers, but object properties ar not linked
  with a vocabulary terms describing semantic meaning of data.

- `5` - all of the lower level requirements are met.

Data with lower maturity level might require more effort to process, but still
data in any form is better, than no data at all. The only thing, that is
required, to specify correct maturity level value if data does not meet the
requirements.

Maturity level is set and can be found in the Catalog.

According to object maturity API responses MUST return following elements:

- **Maturity level 5**
```Code
HTTP/1.1 200 OK
Content-Type: application/ld+json

{
  "@context": "https://data.gov.lt/context/{dataset.id}/{version}.jsonld",
  "@id": "https://data.gov.lt/id/{vocab}/{term}/42",
  "_id": "42",
  "id": 42,
  "title": "Duomuo"
}
```

- **Maturity level 4**
```Code
HTTP/1.1 200 OK
Content-Type: application/ld+json

{

>   "@id": "https://data.gov.lt/id/{vocab}/{term}/42",

Reviewer: `@id` dalis gal būt atsiranda tik nuo 5 brandos lygio, kai yra pateiktas KDSA? Kitu atveju, nežinome kokį URI prefiksą naudoti?

Ir gal būt reikėtu žodžiu užrašyti, kad nuo 4 brandos lygio atsiranda `_id` laukas, kuris visada rodo į objekto identifikatorių.

Ir reikėtu paaiškinti, kad nuo 4 brandos lygio, veiksmai su vienu objektu atliekami nurodant identifikatorių, kaip URL Path elementą. O ne URL Query filtrus.

  "_id": "42",
  "id": 42,
  "title": "Duomuo"
}
```

> - **Maturity level 3**

Reviewer: Siūlyčiau apjungti 1-3 brandos lygį, į vieną grupę, nes iš API perspektyvos ir nežiūrint į semantiką, UDTS veikimas yra vienodas.

Papildomai reikėtu pateikti pavyzdžius ir paaiškinti, kad visi veiksmai, kurie reikalauja identifikatoriaus, veikia tik per URL query, kaip filtras, su explicit action.

Pavyzdžiui:

    Country/:getone/?code=42

Taip pat, jei atliekos rašymo operacijos ir filtruose nurodyti laukai sutampa su pirminiu raktu (`model.ref`), tada draudžiama atlikti veiksmus, kurie keičia daugiau nei vieną objektą.

```Code
HTTP/1.1 200 OK
Content-Type: application/json

{
  "id": 42,
  "title": "Duomuo"
}
```

- **Maturity level 2**
```Code
HTTP/1.1 200 OK
Content-Type: application/json

{
  "code": 1000,
  "title": "Duomuo"
}
```

- **Maturity level 1**
```Code
HTTP/1.1 200 OK
Content-Type: application/json

{
  "title": "Duomuo"
}

> ```

Reviewer: Trūksta nulinio brandos lygio, kuris, kuris neturi jokio identifikatoriaus. Tokiu būdu UDTS API yra išjungiami visi veiksmai "Action", kuriems reikalingas identifikatorius.

Tai reiškia, nėra galimybės gauti vieną pagal id, keisti ar trinti objektą.

enforced by the Agent.

# Tracking data changes

When an object data is created or changed a log is being created and updated to keep track of all of these changes.
This can be retrieved via `change` actions. It can also be used to determine the progression of object changes in time.
This log only store the change that has been done with the object and a revision id `rev` that can be used to retrieve 
entire object at that point in time. Change id `cid` is used to rdetermine the change order for the objects,
it starts at `0` when the objec is created and is increased incrementaly with each change that is done.

Important to note that the log should not be used to retain or retrieve historic object data since it only stores the changes.
Object revision retrieval should be used instead.

To retrieve the latest change of the object `-1` can be used as the `cid`:

```/datasets/gov/rc/ar/uapi/text_with_coordinates/Country/actions/:changes/-1```

Changes actions are implemented at the model and object levels and conforms to [LDES](https://semiceu.github.io/LinkedDataEventStreams/)
For this it uses the [JSON-LD Context](https://www.w3.org/TR/json-ld/#the-context)

> Depending on the size response might be paginated and accessed via query ```?page=n```

Reviewer: ```suggestion
Depending on the size response might be paginated and accessed via query ```?_page=n```
```
